PIP-9:-Topic-Compaction.md

```diff
@@ -0,0 +1,69 @@
+
+Compaction doesn't directly interact with message expiration. Once a topic is compacted, the backlog still exists. However, subscribers with cursors before the compaction horizon will move quickly through the backlog, so it will eventually get cleaned up as there'll be no subscribers.
+
+Open Question: Should the compactor have a subscription so that it can cause messages to be retained until compaction occurs.
+
+## Changes
+
+### The Compactor
+
+The compactor will be a command line application. The application will take the URI as an argument. It will require a valid client configuration.
+
+The compactor worker will be written in a way that it can later be embedded in a daemon. The algorithm will be as described above.
+
+### Client library
+
+It would be better to avoid completely deserializaing and reserializing each message, so that we can sidestep the need for special handling for compression and encryption. To this end, the client needs to provide a "raw" client, that allows the user to access the metadataAndPayload ByteBuf that is currently sent to the client. This ByteBuf is the data as it stored in the managed ledger, and all deserialation of it usually takes place at the client. The compactor will take this ByteBuf, partially deserialize it to get the key from the metadata, and then if it is a candidate for compaction, push it into the compacted topic ledger, prepended with the serialized message id.
+
+Messages read from a compacted topic by a normal client will be no different to messages read from a non-compacted topic, though there will be gaps in the message IDs.
+
+### Broker
+
+The broker requires changes in the PersistentDispatcher and PersistentTopic, and has a new class CompactedTopic.
+
+#### CompactedTopic
+
+The broker has a class CompactedTopic. There are two important calls on this class, #setCompactionSpec and #asyncReadEntries.
+
+setCompactionSpec is called to update the compaction horizon and compacted topic ledger for the topic.
+
+asyncReadEntries is very similar to ManagedCursor, but it instead of reading from a ManagedLedger it reads from the compacted topic ledger.
+
+#### PersistentDispatcher#readMoreEntries
+
+when readMoreEntries is called on the dispatcher, we will check whether cursor is lower than the compaction horizon. If so it reads from the CompactedTopic. Otherwise it reads from the compactor as it does now.
+
+#### PersistentTopic
+
+PersistentTopic will check if compaction is enabled on the topic, and if so, start watching the metadata for changes to the compaction horizon and compacted topic ledger. When it sees an update, the CompactedTopic object will be updated.
+
+
+
+
```

Comment:
I would really like to see an explicit throttle on compaction I/O.

How efficient this would be when multiple topics are getting compacted? That would be almost like backlog draining which currently is I/O bound, when many topics do draining at the same time. Need not directly attempt to do this here,  but I would be interested to know if we can have a layered, bottom up approach
-single scan of entry log, selecting entries for compaction (from one or more ledgers)
-physical compaction of ledger entries belonging to the same ledger 
-logical compaction of the entries from same ledger with same key (as in this PIP)

Reply:
Added a note about throttling. 

Regarding doing this in BK, I'm not sure how much this would win us. Doing logical requires 1 read per entry per ledger, and N writes, where N is the replication factor. Doing compaction on each entrylog would require N reads, and then N writes (in worst case, where all keys are unique).